Linux support

Geenz opened the pull request:
Linux support for the viewer.  Still very much a WIP.

Please target develop-linux if you'd like to contribute to linux support.  We're trying to keep develop more or less clean until we've shipped 2025.08.

akleshchev committed:
Windows build fix

RyeMutt committed:
Remove dead LLPostProcess code with deprecated opengl usages

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Cherrypick 8de8daca601dc85e2b73687856f0a321016b4463 from archive/develop

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Restore llsys linux fixes from archive/develop 57d423745fd1d3d0ea6a0c69b869a20c27e27fc5

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Cherrypick 24854b4dd74edabf67efe533ef191553442a9bad from archive/develop

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix or supress various warnings under GCC builds

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix ODR violation during linux from ld

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix up linux feature table and various small build changes

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Disable viewer-manager in linux builds

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix build errors due to opengl headers and various defines

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix detection of AMD gpu on linux

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Cherrypick 42885c0d23d8efc47ad77e09bb327e4ec6a4fada from archive/develop

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Restore LLMaterialID to POD type

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix boost asio deprecation warnings

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix overflows inserting invalid values into stats map

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Cherrypick LLProcessor cleanup from 57d423745fd1d3d0ea6a0c69b869a20c27e27fc5

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Introduce support for GL_ATI_meminfo for getting gpu memory on linux

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Add debug option to force libvlc plugin on linux

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix numerous compiler warnings from GCC under Ubuntu 22.04

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Start to fix up GHA workflow for linux

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
CMake changes for linux64 support including code from archive/develop
Restore changes for GLIB based dbus support from archive/develop

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix RelWithDebInfo build on mac universal

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix shader cache failing after clearing contents

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix low memory notification appearing incorrectly on non-windows platforms when changing graphics quality levels

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Cherry-pick gstreamer10 from 57d423745fd1d3d0ea6a0c69b869a20c27e27fc5

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix makefile and ninja generators invalidating secondlife-bin target every time cmake is run

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix transparent water shader being stuck as opaque when reset by feature table

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix crash observed during early startup during feature table resets on linux due to LLWorld simpleton not being initialized yet

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Restore linux CEF build support including changes from archive/develop
Fix various files missing linden_common

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Introduce global precompiled header to accelerate build

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Run tests using ctest after all targets are built to reduce build contention
Reuse test runner compilation unit to reduce extra source files compiled by 189
Only build macOS tests for build host architecture that will be running them to reduce wasted CI time

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Reintroduce llappearanceutility to viewer tree from archive/develop

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix shift by negative -1 in llflexibleobject

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix memcpy of 0 being reported as undefined behavior by ubsan

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix invalid cast warning from ubsan

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix leak of LLRender VBCache when reset/shutdown

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix broken mvp modelview hash check in syncMatrices

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix leak in materialmgr reported by asan/leaksan by using unique_ptr

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix uninitialized member variables reported by ubsan

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix leak of boost signals in llmarketplacefunctions

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix leak of LLNotificationSeparator in LLFloaterNotificationsTabbed

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix GL debug logging on linux

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix shader warnings triggering GL errors under linux mesa drivers

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix assert failing debugGL run

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Add support for building with address, undefined, and thread sanitizers on linux

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix tracy memory tracking on linux to not crash mesa GL driver

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Introduce SDL based file picking support

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Introduce SDL3 based windowing backend support and dir picker

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Cleanup up linux wrapper script and fix desktop entry creation

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Modernize viewer alignment macros to use alignas and fix misaligned class issues reported by ubsan/asan

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Update openjpeg to 2.5.4-r1

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix broken locking for decomp data in model preview

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Modernize getAvailableMemoryKB for linux

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix deprecated opengl usage in debug display

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix usage of deprecate opengl defines

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix deprecated opengl mipmap generation usage

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Reduce extra packages installed during autobuild configure

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix deprecated opengl draw commands

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Rework anisotropic filtering options into a dropdown with levels for better performance tuning of low end gpu

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Remove dead linux script

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix cef scaling on linux hidpi displays

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix broken mutex lock on skinned mesh map

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix non-static LLCachedControl

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Add early out for disk cache when app is shutting down to reduce shutdown time

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix multiple unicode file io handling issues with llofstream and llifstream

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Reduce texture cache start up time by reading header entries file in one call instead of many small ones

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix invalid calls to glLineWidth on some linux GL drivers

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix log to file not supporting multiline output

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix threaded GL contexts on linux backend

Enable core profile context support on linux

Signed-off-by: Rye <rye@alchemyviewer.org>

Geenz committed:
Merge branch 'develop-linux' into rye/linux64

RyeMutt committed:
Add support for GLX and EGL to GL extension loader
Fix dullahan wayland support
Fix shader GL_ERROR under GL Core profile in Mesa

Signed-off-by: Rye <rye@alchemyviewer.org>

Geenz committed:
Merge pull request #4827 from RyeMutt/rye/linux64

Linux Support Part 1

RyeMutt committed:
Rework and modernize GL function/extension initialization for better wayland compatability

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Remove dead linux crash logger

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix support for setting thread names on linux and macos

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix support for MSVC toolset v145(VS2026)

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Add basic linux build instructions

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix building with clang on non-darwin platforms

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Add support for tracy logging output

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Update open-libndof package with fixes for SDL3 joystick init

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix registration of secondlife protocol handler on linux

Signed-off-by: Rye <rye@alchemyviewer.org>

Geenz committed:
Merge pull request #4915 from RyeMutt/rye/linux64-2

Linux64 Part 2

Geenz committed:
Merge branch 'develop' into develop-linux

RyeMutt committed:
Add GL constants needed for compatibility with physicsextensions_source package

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Improve fix for line width crashes under certain Mesa GL drivers

We default to and only use aliased lines.

Signed-off-by: Rye <rye@alchemyviewer.org>

Geenz committed:
Merge pull request #4934 from RyeMutt/rye/fixlnxbuild

Add GL constants needed for compatibility with physicsextensions_source package

Geenz committed:
Merge pull request #4935 from RyeMutt/rye/lnxfixlines

Improve fix for line width crashes under certain Mesa GL drivers

niciphilly commented on indra/newview/llfilepicker.cpp:
```diff
@@ -167,7 +167,451 @@ void LLFilePicker::reset()
+        case LLFilePicker::FFLOAD_RAW:
+            filter_vec.push_back({ "RAW files (*.raw)", "raw" });
+            break;
+        case LLFilePicker::FFLOAD_MODEL:
```
I noticed that I can only upload DAE files for models on Linux.
So I think this isn't quite correct. The Windows and Mac implementation also includes `*.gtlf` and `*.glb` files for `LLFilePicker::FFLOAD_MODEL`. I guess it's missing here, too.

niciphilly commented on indra/newview/llfilepicker.cpp:
```diff
@@ -167,7 +167,451 @@ void LLFilePicker::reset()
+        case LLFilePicker::FFLOAD_GLTF:
+            filter_vec.push_back({ "glTF (*.gltf; *.glb)", "gltf;glb" });
+            break;
+        case LLFilePicker::FFLOAD_COLLADA:
```
Couldn't `LLFilePicker::FFLOAD_COLLADA` be dropped completely? I couldn't find any usage besides other `case` blocks anywhere.

niciphilly commented on indra/newview/linux_tools/refresh_desktop_app_entry.sh:
```diff
@@ -13,18 +13,25 @@ function install_desktop_entry()
 Comment=Client for the On-line Virtual World, Second Life\n\
-Exec=${installation_prefix}/secondlife\n\
+Path=${installation_prefix}\n\
+Exec=secondlife\n\
```
This doesn't work on Plasma. It complains `Could not find the program 'secondlife'`. And according to https://specifications.freedesktop.org/desktop-entry/latest/exec-variables.html `Exec` should be an absolute path or the executable needs to be in `$PATH`. `Path` only sets the working directory for the application but isn't used to lookup the executable.

Quinn-Elara replied:
This is incorrect. Because the desktop entry is created from a bash script, the variable (represented by the $ sign) is substituted for the actual (absolute) path.

niciphilly replied:
Yes, it is substituted. But not in the `Exec` line anymore. Right now it results in a desktop file like:
```
Path=/home/nici/installation/dir
Exec=secondlife
```
And this would require that I add `/home/nici/installation/dir` to $PATH. The version before the change was correct:
```
Exec=/home/nici/installation/dir/secondlife
```

Quinn-Elara replied:
In which case, this is likely a difference in how Plasma and Gnome works (and likely different in COSMIC, et al.)
Seemingly, GNOME will interpret the PATH variable as the launch directory (and thus, the working directory), and if what you're saying is correct, then KDE is doing this another way. Easily fixed, will likely PR some point next week.

niciphilly replied:
Yes, it might be something like that. The spec says absolute path or look up with `$PATH`. I guess Gnome does it a bit more relaxed and Plasma sticks to the letters.

RyeMutt committed:
Fix failure to generate build config information correctly under some cmake generators due to CMAKE_BUILD_TYPE only being utilized by non-multiconfig

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix alt cam mouse motion stuttering in variance to framerate under wayland

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix desktop entry lacking absolute path for Exec failing under KDE

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Fix GLTF files missing from model selection under SDL filepicking

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt commented:
Additional fixes in #4954

Geenz committed:
Remove cpuid feature bits verification in llprocessor.cpp (#4959)

RyeMutt committed:
Remove leftover boost aligned_allocator as this is now handled by c++17 aligned allocations

Signed-off-by: Rye <rye@alchemyviewer.org>

AiraYumi committed:
Prevent the creation of temporary objects that will be discarded.

RyeMutt committed:
Optimize post process rendering for gltf material preview

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Migrate sky/cloud rendering to fragment shader and introduce high precision post process option

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Replace dependency on boost thread with std and change HttpService to use a similar thread shutdown mechanism to LLThread

RyeMutt committed:
Fix invalid symbol in gstreamer 1.0 plugin crashing (#5059)

Signed-off-by: Rye <rye@alchemyviewer.org>

marchcat committed:
Merge branch 'release/2026.01' into marchcat/devlinux-2601

# Conflicts:
#	indra/llcommon/llprocessor.cpp
#	indra/llcorehttp/_refcounted.h
#	indra/llcorehttp/_thread.h
#	indra/newview/llfeaturemanager.cpp
#	indra/newview/llviewershadermgr.cpp

marchcat committed:
LLTextBase build fix

marchcat committed:
Merge pull request #5101 from secondlife/marchcat/devlinux-2601

Merge 26.01 into develop-linux

marchcat committed:
Revert "Windows build fix"

This reverts commit 8c292db722d01fcaef95752f93843e8533b05710.

marchcat committed:
Proper Windows build fix (#5105)

RyeMutt committed:
Rework legacy lltest executable into individual integration tests and restore some disabled tests

Removes dead llprimlinkinfo and related test

Remove dead llassetuploadqueue llhttpclient lltimestampcache lltranscode tests

Remove dead mock_http_client

Remove duplicate hex_dump.h

Fix crash in LLDataPacker test

akleshchev committed:
Reapply "Add more functionality to LLFile and cleanup LLAPRFile""

# Conflicts:
#	indra/llcrashlogger/llcrashlock.cpp
#	indra/llrender/llshadermgr.cpp
#	indra/test/CMakeLists.txt

RyeMutt committed:
Restore multiple disabled tests in llcharacter and newview

Clean up dead test files from llcommon and newview

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Prune files from macOS GHA runner to fix build failure from disk space exhaustion

Hecklezz committed:
Cleaned up unnecessary fake bool usage where appropriate

Signed-off-by: Hecklezz <tj8@live.com.au>

RyeMutt committed:
Clean up dead partsyspacket code

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Clean up dead LLStreamQueue code and related tests

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Clean up dead LLBase32 code

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Clean up dead LLSphere code and related tests

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Clean up dead TeamCity build output and work arounds from tests

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Remove dead LLLooksHistoryPanel LLPanelIMControlPanel and LLToolView

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Clean up dead legacy headers and compiler work arounds

Clean up dead macos files

Clean up dead windows build files

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Split Viewer and Test builds into separate jobs in GHA to reduce potential for disk space exhaustion

Signed-off-by: Rye <rye@alchemyviewer.org>

akleshchev committed:
#4899 Adjustments per rae's review

akleshchev committed:
#4899 Add unicode test case

akleshchev committed:
#4899 Add unicode test case #2

akleshchev committed:
#4899 Fix erros related to negative file size

RyeMutt committed:
Replace remaining boost::filesystem usage with std::filesystem

Signed-off-by: Rye <rye@alchemyviewer.org>

AndrewMeadows committed:
make LLCore::HttpStatus type enums constexpr

RyeMutt committed:
Clean up LLFile and LLFilesystem operations to reduce temporaries allocations and utf8->utf16 conversions
Fix LLFile::tmpdir failing when temp directory is located in a unicode path on windows
Add Tracy profiler tagging to LLFile that's disabled by default
Replace LLFile::utf8StringToWstring with our fsyspath std::filesystem::path adapter
Introduce std::filesystem::path to llifstream/llofstream wrappers
Add move operators to fsyspath to match copy operator
Restore LLUniqueFile RAII class for direct c file io wrapping and reducing merge conflicts with downstream code that may depend upon it

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Reduce temporary allocations and repeated utf8->utf16 in LLTextureCache with new LLFile IO code

Add non-allocating to_chars and to_wchars string creation functions to LLUUID

Signed-off-by: Rye <rye@alchemyviewer.org>

RyeMutt committed:
Address feedback from copilot

Signed-off-by: Rye <rye@alchemyviewer.org>

akleshchev committed:
#4899 Fix permissions flags on files that shouldn't be executable

akleshchev committed:
#4899 Fix permissions flags on files that shouldn't be executable #2

Hecklezz committed:
Corrected file permissions, only permitting execution on .sh scripts and .py scripts with a shebang

Signed-off-by: Hecklezz <tj8@live.com.au>

RyeMutt committed:
Merge remote-tracking branch 'origin/develop' into develop-linux

akleshchev committed:
Merge pull request #5227 from RyeMutt/rye/devmerge

Sync develop-linux with develop

FelixWolf committed:
Add additional screenshot resolution options to inventory and local panels (#5292, #5295)

FelixWolf committed:
Refactor SDL references to remove version specificity and improve clarity (#5285)

RyeMutt committed:
Merge remote-tracking branch 'origin/develop' into develop-linux

akleshchev committed:
Merge develop into develop-linux

Merge pull request #5316 from RyeMutt/rye/devlinuxupdate

RyeMutt committed:
secondlife/viewer#4958 Fix model upload options for SDL file picker

tapple committed:
Some notes in linux build instructions (#5388)

* Some notes in linux build instructions

* note clang stability

i-doll committed:
Fix locale thousands separator producing '?' on Linux with multibyte UTF-8 separators

getThousandsSeparator() and getMonetaryThousandsSeparator() extracted only
the first byte of localeconv()->thousands_sep, truncating multibyte UTF-8
separator characters (e.g. non-breaking space U+00A0). The incomplete byte
was then embedded via %c format specifier, producing invalid UTF-8 rendered
as '?' in notifications like avatar complexity.

Return the full separator string and use string concatenation instead of %c.

Hecklezz committed:
Fix LLTextureCacheRemoteWorker::doRead malloc size error
